src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.VariableIdentifier.cs

```diff
@@ -26,26 +27,60 @@ public bool Exists
             {
                 get { return (object)Symbol != null; }
             }
+
             public override int GetHashCode()
             {
-                return Roslyn.Utilities.Hash.Combine(Symbol?.OriginalDefinition, ContainingSlot);
+                Debug.Assert(Exists);
+
+                int currentKey = ContainingSlot;
+                // MemberIndexOpt, if available, is a cheap approach to comparing relative members.
+                int? thisIndex = Symbol.MemberIndexOpt;
+                return thisIndex.HasValue ?
+                    Hash.Combine(Symbol.OriginalDefinition, currentKey) :
+                    Hash.Combine(thisIndex.GetValueOrDefault(), currentKey);
             }
+
             public bool Equals(VariableIdentifier other)
             {
-                return ((object)Symbol == null ? (object)other.Symbol == null : Symbol.OriginalDefinition.Equals(other.Symbol.OriginalDefinition)) && ContainingSlot == other.ContainingSlot;
+                Debug.Assert(Exists);
+                Debug.Assert(other.Exists);
+
+                if (ContainingSlot != other.ContainingSlot)
+                {
+                    return false;
+                }
+
+                // MemberIndexOpt, if available, is a cheap approach to comparing relative members.
+                int? thisIndex = Symbol.MemberIndexOpt;
+                int? otherIndex = other.Symbol.MemberIndexOpt;
+                if (thisIndex != otherIndex)
+                {
+                    return false;
+                }
+
+                if (thisIndex.HasValue)
```

Comment:
>if (thisIndex.HasValue) [](start = 16, length = 23)

Consider replacing this `if` and the preceding one with:
```C#
if (this.Index.HasValue)
{
    return thisIndex != otherIndex;
}
```

Reply:
That would not compare the indices when `thisIndex.HasValue == false` and `otherIndex.HasValue == true`.

---
In reply to: [259137280](https://github.com/dotnet/roslyn/pull/33562#discussion_r259137280) [](ancestors = 259137280)